hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordScanner.java

```diff
@@ -113,8 +120,11 @@ protected AbstractHoodieLogRecordScanner(FileSystem fs, String basePath, List<St
     this.readerSchema = readerSchema;
     this.latestInstantTime = latestInstantTime;
     this.hoodieTableMetaClient = HoodieTableMetaClient.builder().setConf(fs.getConf()).setBasePath(basePath).build();
+    this.orderingField = this.hoodieTableMetaClient.getTableConfig().getPreCombineField();
     // load class from the payload fully qualified class name
     this.payloadClassFQN = this.hoodieTableMetaClient.getTableConfig().getPayloadClass();
+    // only the OverwriteWithLatestAvroPayload needs combining by orderingVal
+    this.naturalOrder = !OverwriteWithLatestAvroPayload.class.isAssignableFrom(ReflectionUtils.getClass(this.payloadClassFQN));
```

Comment:
Couple of points to ponder
- We should try to be agnostic to payload impl's in this class. Should not leak OverwriteWithLatestAvroPayload
- One another option is to add another constructor which takes in orderingField (instead of ordering Val) to the payload class. Basically this will move the line 318 to within Payload constructor itself. Basically one constructor will be used in the direct write path from user (ordering val) and another constructor will be used when reading a record already persisted.

Reply:
Agree, would add a new constructor.

Reply:
One problem is that may pay load impls implement the interface `HoodieRecordPayload` directly, and there is no need to do preCombine for them.